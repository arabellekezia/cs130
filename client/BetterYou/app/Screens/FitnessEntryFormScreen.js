import React, { useState } from "react";

import { StyleSheet, Text, View } from "react-native";
import DropDownPicker from "react-native-dropdown-picker";

import colors from "../config/colors";
import Screen from "../components/Screen";
import AppText from "../components/AppText";
import AppTextInput from "../components/AppTextInput";
import TextButton from "../components/TextButton";
import ErrorMessage from "../components/ErrorMessage";
import { ScrollView } from "react-native-gesture-handler";

function isValidInput(category, activeTime, setError) {
  const validCategory = isValidCategory(category);
  const validActiveTime = isValidActiveTime(activeTime);
  setError({ category: !validCategory, activeTime: !validActiveTime });
  return validCategory && validActiveTime;
}

function isNumber(value) {
  return typeof value === "number" && isFinite(value);
}

function isValidActiveTime(activeTime) {
  return isNumber(activeTime) && activeTime > 0;
}

function isValidCategory(category) {
  return category.length > 0;
}

function FitnessEntryFormScreen({ navigation, route }) {
  const initTime = route.params ? route.params.elapsedTime / 1000 / 60 : 0;

  const [category, setCategory] = useState("");
  const [activeTime, setActiveTime] = useState(initTime);
  const [err, setError] = useState({ category: false, activeTime: false });

  const activities = [
    { label: "Cycling", value: "Bicycling: 12-13.9 mph" },
    { label: "Hiking", value: "Hiking: cross country" },
    { label: "Jogging", value: "Running: 6 min/mile" },
    { label: "Sprinting", value: "Running: 10 min/mile" },
    { label: "Swimming", value: "Swimming: laps, vigorous" },
    { label: "Walking", value: "Walk: 15 min/mile" },
    { label: "Weightlifting", value: "Weightlifting: general" },
  ];

  function submit() {
    if (!isValidInput(category, activeTime, setError)) {
      console.log(err);
      return;
    }
    console.log(category, activeTime);
    navigation.popToTop();
  }

  function displayErrorMessage(error) {
    if (error.category) {
      return <ErrorMessage message="Must select a category." />;
    }
    if (error.activeTime) {
      return <ErrorMessage message="Must input a number greater than 0." />;
    }
  }

  return (
    <Screen style={styles.container}>
      <ScrollView style={{ padding: 20 }} keyboardShouldPersistTaps="handled">
        <AppText style={styles.text}>What activity did you do?</AppText>

        <DropDownPicker
          defaultNull
          placeholder="Select an item"
          items={activities}
          containerStyle={{ height: 40 }}
          style={styles.dropdownInner}
          labelStyle={{ fontSize: 18, color: colors.dark }}
          containerStyle={styles.dropDownContainer}
          placeholderStyle={styles.dropDownPlaceholder}
          dropDownStyle={styles.dropDownList}
          onChangeItem={(item) => {
            setCategory(item.value);
            setError({ category: false, activeTime: false });
          }}
        />

        <AppText style={styles.text}>How many minutes were you active?</AppText>
        <AppTextInput
          placeholder="30 minutes"
          value={activeTime > 0 ? activeTime.toString() : ""}
          keyboardType="numeric"
          onChangeText={(time) => {
            setActiveTime(Number.parseFloat(time));
            setError({ category: false, activeTime: false });
          }}
        />
        {displayErrorMessage(err)}
        <TextButton
          style={styles.button}
          name="Submit"
          onPress={() => submit()}
        />
      </ScrollView>
    </Screen>
  );
}

const styles = StyleSheet.create({
  button: {
    marginVertical: "5%",
  },
  container: {
    flex: 1,
    justifyContent: "center",
    backgroundColor: colors.background,
  },
  dropDownContainer: {
    height: 50,
    marginTop: 10,
  },
  dropdownInner: {
    backgroundColor: "#efefef",
  },
  dropDownLabels: {
    fontSize: 18,
    color: colors.dark,
  },
  dropDownList: {
    backgroundColor: "#fafafa",
  },
  dropDownPlaceholder: {
    color: "#7e7e7e",
  },
  text: {
    marginTop: "5%",
    fontSize: 16,
  },
});

export default FitnessEntryFormScreen;
