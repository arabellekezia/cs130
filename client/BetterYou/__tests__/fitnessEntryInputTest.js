import {render, cleanup, fireEvent} from '@testing-library/react-native';
import React from 'react';
import FitnessEntryFormScreen from "../app/Screens/FitnessEntryFormScreen"

describe('Fitness entry input form', () => {
  let wrapper;
  beforeEach(() => {
      wrapper = render(<FitnessEntryFormScreen route={{}} />);
    });
  afterEach(() => {
      cleanup();
      wrapper = null;
    });
  it('renders correctly', () => {
      expect(wrapper).toMatchSnapshot();
  });

  it("Shows error message if no category is selected", () => {
    const {getByText, queryByTestId} = wrapper;
    
    const submitButton = getByText("Submit");
    fireEvent.press(submitButton);
  
    const validationError = queryByTestId("error-message");
    expect(validationError).toBeTruthy();
    expect(validationError.props.children)
        .toBe("Must select a category.");
  });

  it("Shows error message if no active time minutes is inputted.", () => {
    const {getByText, queryByTestId} = wrapper;

    const cyclingOption = getByText("Cycling");
    fireEvent.press(cyclingOption);

    const submitButton = getByText("Submit");
    fireEvent.press(submitButton);
  
    const validationError = queryByTestId("error-message");
    expect(validationError).toBeTruthy();
    expect(validationError.props.children)
        .toBe("Must input active time greater than 0.");
  });

  it("Shows error message if no weight is inputted", () => {
    const {getByText, queryByTestId, getByPlaceholderText} = wrapper;

    const cyclingOption = getByText("Cycling");
    fireEvent.press(cyclingOption);

    const activeMinutesTextInput = getByPlaceholderText("30 minutes"); 
    const enteredMinutes = "25,";
    fireEvent(activeMinutesTextInput, 'onChangeText', enteredMinutes);

    const submitButton = getByText("Submit");
    fireEvent.press(submitButton);
  
    const validationError = queryByTestId("error-message");
    expect(validationError).toBeTruthy();
    expect(validationError.props.children)
        .toBe("Must input a weight greater than 0.");
  });

  it("No error is displayed with all proper inputs.", () => {
    const {getByText, queryByTestId, getByPlaceholderText} = wrapper;

    const cyclingOption = getByText("Cycling");
    fireEvent.press(cyclingOption);

    const activeMinutesTextInput = getByPlaceholderText("30 minutes"); 
    const enteredMinutes = "25,";
    fireEvent(activeMinutesTextInput, 'onChangeText', enteredMinutes);

    const weightTextInput = getByPlaceholderText("140 pounds"); 
    const enteredWeight = "140";
    fireEvent(weightTextInput, 'onChangeText', enteredWeight);

    const submitButton = getByText("Submit");
    fireEvent.press(submitButton);
  
    const validationError = queryByTestId("error-message");
    expect(validationError).toBe(null);
  });
});

