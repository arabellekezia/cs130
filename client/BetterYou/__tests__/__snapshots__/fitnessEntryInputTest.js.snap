// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fitness entry input form renders correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
        "flex": 1,
        "justifyContent": "center",
      },
      Object {
        "paddingTop": 44,
      },
    ]
  }
>
  <RCTScrollView
    collapsable={false}
    keyboardShouldPersistTaps="handled"
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    style={
      Object {
        "padding": 20,
      }
    }
  >
    <View>
      <Text
        style={
          Array [
            Object {},
            Object {
              "fontSize": 16,
              "marginTop": "5%",
            },
          ]
        }
      >
        What activity did you do?
      </Text>
      <View
        style={
          Array [
            Object {
              "height": 50,
              "marginTop": 10,
            },
            Object {
              "zIndex": 5000,
            },
          ]
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onLayout={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#efefef",
              "borderBottomLeftRadius": 5,
              "borderBottomRightRadius": 5,
              "borderColor": "#dfdfdf",
              "borderTopLeftRadius": 5,
              "borderTopRightRadius": 5,
              "borderWidth": 1,
              "flex": 1,
              "flexDirection": "row",
              "opacity": 1,
              "paddingHorizontal": 10,
              "paddingVertical": 5,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#3d3d3d",
                  "fontSize": 18,
                },
                Object {
                  "color": "#7e7e7e",
                },
                Object {
                  "flex": 1,
                  "opacity": 1,
                },
                Object {
                  "marginLeft": 5,
                  "marginRight": 0,
                },
                false,
                Object {
                  "marginLeft": 5,
                },
              ]
            }
          >
            Select an item
          </Text>
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "borderBottomLeftRadius": 0,
                  "borderTopLeftRadius": 0,
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "paddingVertical": 8,
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {},
                  Object {
                    "opacity": 1,
                  },
                ]
              }
            >
              <Text />
            </View>
          </View>
        </View>
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#fff",
                "borderBottomLeftRadius": 5,
                "borderBottomRightRadius": 5,
                "borderColor": "#dfdfdf",
                "borderTopLeftRadius": 5,
                "borderTopRightRadius": 5,
                "borderWidth": 1,
                "paddingHorizontal": 10,
                "paddingVertical": 5,
              },
              Object {
                "alignItems": "center",
                "borderTopLeftRadius": 0,
                "borderTopRightRadius": 0,
                "justifyContent": "center",
                "position": "absolute",
                "width": "100%",
              },
              Object {
                "backgroundColor": "#fafafa",
              },
              Object {
                "borderWidth": 0,
                "display": "none",
                "position": "relative",
              },
              Object {
                "maxHeight": 150,
                "top": undefined,
                "zIndex": 5000,
              },
            ]
          }
        >
          <RCTScrollView
            nestedScrollEnabled={true}
            style={
              Object {
                "width": "100%",
              }
            }
          >
            <View>
              <View
                onLayout={[Function]}
              >
                <View
                  accessible={true}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "center",
                      "opacity": 1,
                      "paddingVertical": 8,
                      "width": "100%",
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "alignContent": "center",
                        "flexDirection": "row",
                      }
                    }
                  >
                    <Text
                      style={
                        Array [
                          Object {
                            "color": "#3d3d3d",
                            "fontSize": 18,
                          },
                          false,
                          Object {},
                        ]
                      }
                    >
                      Cycling
                    </Text>
                  </View>
                </View>
              </View>
              <View
                onLayout={[Function]}
              >
                <View
                  accessible={true}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "center",
                      "opacity": 1,
                      "paddingVertical": 8,
                      "width": "100%",
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "alignContent": "center",
                        "flexDirection": "row",
                      }
                    }
                  >
                    <Text
                      style={
                        Array [
                          Object {
                            "color": "#3d3d3d",
                            "fontSize": 18,
                          },
                          false,
                          Object {},
                        ]
                      }
                    >
                      Hiking
                    </Text>
                  </View>
                </View>
              </View>
              <View
                onLayout={[Function]}
              >
                <View
                  accessible={true}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "center",
                      "opacity": 1,
                      "paddingVertical": 8,
                      "width": "100%",
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "alignContent": "center",
                        "flexDirection": "row",
                      }
                    }
                  >
                    <Text
                      style={
                        Array [
                          Object {
                            "color": "#3d3d3d",
                            "fontSize": 18,
                          },
                          false,
                          Object {},
                        ]
                      }
                    >
                      Jogging
                    </Text>
                  </View>
                </View>
              </View>
              <View
                onLayout={[Function]}
              >
                <View
                  accessible={true}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "center",
                      "opacity": 1,
                      "paddingVertical": 8,
                      "width": "100%",
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "alignContent": "center",
                        "flexDirection": "row",
                      }
                    }
                  >
                    <Text
                      style={
                        Array [
                          Object {
                            "color": "#3d3d3d",
                            "fontSize": 18,
                          },
                          false,
                          Object {},
                        ]
                      }
                    >
                      Sprinting
                    </Text>
                  </View>
                </View>
              </View>
              <View
                onLayout={[Function]}
              >
                <View
                  accessible={true}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "center",
                      "opacity": 1,
                      "paddingVertical": 8,
                      "width": "100%",
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "alignContent": "center",
                        "flexDirection": "row",
                      }
                    }
                  >
                    <Text
                      style={
                        Array [
                          Object {
                            "color": "#3d3d3d",
                            "fontSize": 18,
                          },
                          false,
                          Object {},
                        ]
                      }
                    >
                      Swimming
                    </Text>
                  </View>
                </View>
              </View>
              <View
                onLayout={[Function]}
              >
                <View
                  accessible={true}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "center",
                      "opacity": 1,
                      "paddingVertical": 8,
                      "width": "100%",
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "alignContent": "center",
                        "flexDirection": "row",
                      }
                    }
                  >
                    <Text
                      style={
                        Array [
                          Object {
                            "color": "#3d3d3d",
                            "fontSize": 18,
                          },
                          false,
                          Object {},
                        ]
                      }
                    >
                      Walking
                    </Text>
                  </View>
                </View>
              </View>
              <View
                onLayout={[Function]}
              >
                <View
                  accessible={true}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "center",
                      "opacity": 1,
                      "paddingVertical": 8,
                      "width": "100%",
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "alignContent": "center",
                        "flexDirection": "row",
                      }
                    }
                  >
                    <Text
                      style={
                        Array [
                          Object {
                            "color": "#3d3d3d",
                            "fontSize": 18,
                          },
                          false,
                          Object {},
                        ]
                      }
                    >
                      Weightlifting
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </RCTScrollView>
        </View>
      </View>
      <Text
        style={
          Array [
            Object {},
            Object {
              "fontSize": 16,
              "marginTop": "5%",
            },
          ]
        }
      >
        How many minutes were you active?
      </Text>
      <View
        style={
          Object {
            "backgroundColor": "#efefef",
            "borderRadius": 10,
            "flexDirection": "row",
            "marginVertical": 10,
            "padding": 15,
            "width": "100%",
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          keyboardType="numeric"
          maxLength={25}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder="30 minutes"
          rejectResponderTermination={true}
          style={
            Object {
              "fontSize": 18,
            }
          }
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
      <Text
        style={
          Array [
            Object {},
            Object {
              "fontSize": 16,
              "marginTop": "5%",
            },
          ]
        }
      >
        How much do you currently weigh?
      </Text>
      <View
        style={
          Object {
            "backgroundColor": "#efefef",
            "borderRadius": 10,
            "flexDirection": "row",
            "marginVertical": 10,
            "padding": 15,
            "width": "100%",
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          keyboardType="numeric"
          maxLength={25}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder="140 pounds"
          rejectResponderTermination={true}
          style={
            Object {
              "fontSize": 18,
            }
          }
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#007AFF",
              "borderRadius": 10,
              "justifyContent": "center",
              "marginVertical": "5%",
              "minWidth": 110,
              "paddingHorizontal": 30,
              "paddingVertical": 12,
            }
          }
        >
          <Text
            style={
              Array [
                Object {},
                Object {
                  "color": "white",
                  "fontSize": 18,
                  "fontWeight": "600",
                },
              ]
            }
          >
            Submit
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
